File: python/150_逆波兰表达式求值_medium.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        if not tokens:
            return 0
        f = {
            '+': add,
            '-': sub,
            '*': mul,
            '/': lambda x, y : int(x/y)
        }
        data = list()
        for token in tokens:
            try:
                num = int(token)
            except:
                x = data.pop()
                y = data.pop()
                num = f[token](y, x)
            finally:
                data.append(num)
        return data[0]

File: cpp/150_逆波兰表达式求值_medium.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        if (tokens.empty())
            return 0;
        stack<int> data;
        // stack.push();
        for (int i=0; i<tokens.size(); i++) {
            if (tokens[i] == "+" || tokens[i] == "-" || tokens[i] == "*" || tokens[i] == "/") {
                int x = data.top();
                data.pop();
                int y = data.top();
                data.pop();
                if (tokens[i] == "+")
                    data.push(x+y);
                if (tokens[i] == "-")
                    data.push(y-x);
                if (tokens[i] == "*")
                    data.push(x*y);
                if (tokens[i] == "/")
                    data.push(y/x);
            }
            else 
                data.push(atoi(tokens[i].c_str()));
        }
        return data.top();

    }
};